### ISSUE TITLE
Timeout Error and setup issues

### ISSUE THREAD

**Issue by basiszwo** [wearehanno/oskar#15]
I am trying to setup oskar but I am struggling with it. I use the deploy heroku button in the readme and setup itself goes fine. After it I can visit the dashboard which is empty.

First I was wondering that oskar doesn't immediately asks me about my feelings. When will he ask first time?

After just giving him a number I cannot access the dashboard anymore. I can see a timeout error when checking the heroku logs. Inspecting mongolab shows me that there's no collection created.

Do you have any idea what went wrong? I really like to try oskar.

**Comment by jonlay** [comment]
Hey @basiszwo! Sorry about this: we're seeing a few problems at our end too, especially with initial onboarding, which throws up some errors sometimes.

@marcelkalveram is tied up on a sprint elsewhere at the moment but I know he plans to try and dig into this in the next week or so. We'll update on here when we can manage to get Oskar chatting a little bit more :smile:

**Comment by basiszwo** [comment]
Thanks for the update. 
I was basically just wondering why I get these timeouts and why the database is not set up properly. 

Maybe there's a manual command so get it up and running?

**Comment by marcelkalveram** [comment]
Hey @basiszwo, thanks for trying out Oskar and for the feedback.

Can you check if there is an environment variable "MONGOLAB_URI" in your Heroku config:
https://www.dropbox.com/s/3g5xt9e9vuj40ty/Screenshot%202015-07-23%2017.58.35.png?dl=0

Just want to make sure that this is not causing the issue. Thanks

**Comment by basiszwo** [comment]
yes, basically the setting looked good and I could hop over to mongolab.

**Comment by marcelkalveram** [comment]
If you want to debug yourself, you can use `foreman start web` on your local machine and use a local database together with your slack account and see if it writes data to it.

I've just given it a try myself and it works, using the env file to define variables for my local environment. Only thing to have in mind is to add a database name to the mongo url:
MONGOLAB_URI=mongodb://127.0.0.1:27017/oskar

Maybe you can give this is a try and see if it saves any data?
Can you also share a screenshot of the console output in the Oskar dashboard?

Thanks!

**Comment by kardotim** [comment]
Same issue here, after entering login credentials on the app's dashboard it just keeps loading and the Heroku log shows a 'request timeout' error :sob:

**Comment by marcelkalveram** [comment]
Looks like there was a timeout issue because Oskar tried to retrieve the status for all team members, even for those that haven't submitted their status yet. Right now it should fall back to an empty state. I just commited a fix (05dab4f). Please let me know if that works. 

Can you also tell me what's in your database? Does it create a `users` collection and is there anything in there?

Thanks!

**Comment by kardotim** [comment]
@marcelkalveram hmm before I was able to see the Oskar front page on our Heroku app, but now I get application error right away.

> Can you also tell me what's in your database? Does it create a users collection and is there anything in there?

Yes it does, our users are in there. We also have set up a database user and added the correct var to the Heroku app. I'll ask our dev to double check to see if it's something on our end though.

**Comment by marcelkalveram** [comment]
That's weird, what are the Heroku logs saying?

I've also just commited a little update that prevents the .js files from being commited, so they can be built on the server from the CoffeeScript source. 

All you have to run is this in order to install the buildpack on Heroku:
`heroku create --buildpack https://github.com/aergonaut/heroku-buildpack-coffeescript.git`

Let me know if that works and if they dashboard is still not showing up.

**Comment by marcelkalveram** [comment]
Ok I just submitted a couple of fixes (19de2e1f6092dbe69fb9bbc77b3adcc24a48af36)

The dashboard was timing out when only one team member had submitted a status. I also found an error in the onboarding process where it didn't read the name of the user because I had previously removed that condition from the `composeMessage` function. I just added it back.

One thing you still need to make sure is that each member has a filled out the "First name" field in their profiles, but I'll create a fallback for that as well, probably tomorrow.

**Comment by johanbrook** [comment]
Hi! I'm @kardotim's fellow developer. After running the buildpack command in our local repo, I get this when trying to deploy:

```
-----> Fetching custom git buildpack... done
-----> Coffeescript app detected
-----> Resolving engine versions
       WARNING: No version of Node.js specified in package.json, see:
       https://devcenter.heroku.com/articles/nodejs-versions
       WARNING: The default version of Node.js and npm on Heroku will begin
       tracking the latest stable release starting September 1, 2012.
       Using Node.js version: 0.4.7
       Using npm version: 1.0.106
-----> Fetching Node.js binaries
-----> Vendoring node into slug
-----> Installing dependencies with npm
       /tmp/node-node-Q6MG/bin/node: error while loading shared libraries: libssl.so.0.9.8: cannot open shared object file: No such file or directory
 !     Failed to install --production dependencies with npm
 !     Push rejected, failed to compile Coffeescript app
```

I find it odd that it complains of Node version, it's specified here: https://github.com/lookback/oskar/blob/master/package.json#L54.

**Comment by marcelkalveram** [comment]
I just ran into the same issue. Sorry for that, but it looks like the CoffeeScript buildpack for Heroku is outdated and causing these errors.

I've removed the buildpack isntructions from the readme file. If you've installed it before, make sure to remove the buildpack URL from your Heroku env variables.

I've removed /target from the gitignore file, as I can't think of any better solution right now than compiling the stuff locally and then pushing it to the repo.

Hope that helps!

**Comment by dirkkelly** [comment]
Thanks for creating Oskar, it's seeming like a really useful bot.

Unfortunately @marcelkalveram fwiw we have just started experience similar timeout issues.

A few people were asked for their status, and provided feedback. For about an hour the dashboard worked.

Now though we're getting a h12 timeout whenever we try to access the dashboard.

```
2015-08-21T18:18:30.191415+00:00 app[web.1]: request user feedback: U04J8FPME
2015-08-21T18:18:30.191522+00:00 app[web.1]: onboarding status U04J8FPME: 0
2015-08-21T18:18:36.001556+00:00 app[web.1]: onboarding status U024H5ZKY: 0
2015-08-21T18:18:55.216524+00:00 heroku[router]: at=error code=H12 desc="Request timeout" method=GET path="/dashboard" dyno=web.1 connect=1ms service=30006ms status=503 bytes=0
2015-08-21T18:19:01.715668+00:00 app[web.1]: onboarding status U04J8FPME: 1
2015-08-21T18:19:11.217092+00:00 app[web.1]: onboarding status U024H5ZKY: 0
2015-08-21T18:19:11.783041+00:00 app[web.1]: onboarding status U04J8FPME: 2
```

I've pushed the latest master branch of the project to heroku, this included https://github.com/wearehanno/oskar/commit/19de2e1f6092dbe69fb9bbc77b3adcc24a48af36

Still receiving timeouts, let me know if there's better debug information I can provide.

**Comment by jonlay** [comment]
Hey there @dirkkelly.

Unfortunately @marcelkalveram is on a 2 week break right now and is totally offline while he's away, so it'll take a little while to reply to you.

I'm limited in what I can do to debug remotely as I'm not as good a developer as Marcel, and am also less familiar with the Oskar codebase. But if you feel up to it, I might suggest that you fork our codebase and have a little look to see if you can the dashboard problem yourself.

Our own team has only ~10 users on it and when we were testing Oskar originally, we found that bigger teams threw up some problems we'd not anticipated (stuff like missing usernames or profile names, and just weird use cases). Sometimes this made the dashboard a little unhappy when we'd forgotten to factor this in.

My best guess is that somewhere on this page (https://github.com/wearehanno/oskar/blob/master/src/views/pages/dashboard.ejs), the conditionals will need to be a little more specific because there'll be something like a missing piece of user information. Perhaps Oskar worked okay at first, and then someone on your team logged in who had a slightly strange username or something, and this is what caused the dashboard to choke.

Sorry to not be able to give a more helpful answer. If you do find a fix, I'd love to hear it and we can happily bring a PR into the repo.

**Comment by pgrujic** [comment]
I am getting the same issue with timeouts, mongodb is collecting data and oskar goes in and out of slack but can't access dashboard I get timeout.

**Comment by marcelkalveram** [comment]
Hey @pgrujic I've been on vacation but will try to have a look at this issue until the end of the week. Thanks for your patience.

**Comment by kardotim** [comment]
Hope you had a good trip @marcelkalveram - looking forward to your commits <3

**Comment by kardotim** [comment]
@marcelkalveram probably no high prio but are you planning to look into this soon?

**Comment by gordonjl** [comment]
I'd love to be able to recommend oskar to my teams, but the timeout issue has been a real beast, and I simply can't run the app. I know contributing a slack bot on open source isn't necessarily Hanno's main business objective, but it would certainly put your organization in a good light if the startup process was fixed. 

I'll dig into the code to see where I can help. If I find a fix, I'll send a PR your way.

**Comment by marcelkalveram** [comment]
Hey there, sorry for the delay. I've been unable to work last week due to medical issues, and was mainly busy on client projects before that. I'll look into your issues this week and get back to you if I have any questions.

**Comment by marcelkalveram** [comment]
I've just added Oskar from scratch on a separate client account, and I might have found a potential issue.
When adding the disabled users and disabled channels in Heroku, do you add double quotations marks around the values? (see screenshot).

![screenshot-2015-10-27-11 15 31](https://cloud.githubusercontent.com/assets/461160/10755222/9749d706-7c9c-11e5-845c-fea7ca91b2ea.jpg)

**Comment by gordonjl** [comment]
I do add double quotes around the disabled channels and users. I do it via the heroku settings page, and it definitely is reading them.

I have disabled everyone on my team (24 members) except myself. Oskar asked me how I was doing today--it also showed the dashboard page and my personal page fine yesterday. This morning it asked my status, but now I get an Application Error when trying to view the dashboard. It's different than the time out; it happens instantly. The Heroku log states the following:

    2015-10-27T12:50:48.969355+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/dashboard" host=pww-slack-oskar.herokuapp.com request_id=48114b79-b5ac-43f0-a258-3958b981d263 fwd="66.231.213.93" dyno= connect= service= status=503 bytes=

Is there any way that I can view more detailed logging information?

Thanks for looking into this. I appreciate it.

**Comment by marcelkalveram** [comment]
That's a common issue with Heroku apps and I should have mentioned it in the readme file, but one of the first things I did was setting up [KeepAwake](https://keep-awake.herokuapp.com/) so that Oskar never falls asleep.

**Comment by marcelkalveram** [comment]
I've also just fixed another issue (c0dab6a1b586bc037178104e6bf182c23c37b9b8) where sorting wasn't possible if at least one user didn't have a status assigned. That was causing timeouts as well.

**Comment by gordonjl** [comment]
Thanks, @marcelkalveram ! re the keepawake: I shelled out the $7 for a hobby version of the heroku dyno. According to the docs, hobby dynos don't sleep. I'll pull in the new code to my app and see how things go.
